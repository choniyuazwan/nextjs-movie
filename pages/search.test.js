import '@testing-library/jest-dom';
import { render, screen, cleanup, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Router from 'next/router';
import Search from '../pages/search';

jest.mock('next/router', ()=> ({push: jest.fn()}));

describe('Search page', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      query: {page: 3, searchkey: 'a'},
      total_pages: 10,
    };
  });

  afterEach(() => {
    mockProps = {};
    cleanup();
  });

  test('should render correctly', () => {
    const { container } = render(<Search {...mockProps} />);
    const main = screen.getByRole('main');
    expect(main).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('should call proper url when user click number page', async () => {
    const { getByTestId } = render(<Search {...mockProps} />);
    await waitFor(() => {
      userEvent.click(getByTestId('button-page-3'));
      expect(Router.push).toHaveBeenCalledWith('/search/?searchkey=a&page=3');
    });
  });

  test('should call proper url when user click first page', async () => {
    const { getByTestId } = render(<Search {...mockProps} />);
    await waitFor(() => {
      userEvent.click(getByTestId('first-page-button'));
      expect(Router.push).toHaveBeenCalledWith('/search/?searchkey=a&page=1');
    });
  });

  test('should call proper url when user click prev page', async () => {
    const { getByTestId } = render(<Search {...mockProps} />);
    await waitFor(() => {
      userEvent.click(getByTestId('button-prev-page'));
      expect(Router.push).toHaveBeenCalledWith('/search/?searchkey=a&page=2');
    });
  });

  test('should call proper url when user click next page', async () => {
    const { getByTestId } = render(<Search {...mockProps} />);
    await waitFor(() => {
      userEvent.click(getByTestId('button-next-page'));
      expect(Router.push).toHaveBeenCalledWith('/search/?searchkey=a&page=4');
    });
  });

  test('should call proper url when user click last page', async () => {
    const { getByTestId } = render(<Search {...mockProps} />);
    await waitFor(() => {
      userEvent.click(getByTestId('button-last-page'));
      expect(Router.push).toHaveBeenCalledWith('/search/?searchkey=a&page=10');
    });
  });

  test('should call proper url when user searching', async () => {
    const { getByTestId } = render(<Search {...mockProps} />);
    await waitFor(() => {
      userEvent.click(getByTestId('search-button'));
      expect(Router.push).toHaveBeenCalledWith('/');
    });
  });
});
