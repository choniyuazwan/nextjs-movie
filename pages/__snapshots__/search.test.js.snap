// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search page should render correctly 1`] = `
<div>
  <main>
    <nav
      class="navbar navbar-expand navbar-light bg-light"
    >
      <div
        class="navbar-collapse collapse"
        id="basic-navbar-nav"
      >
        <div
          class="mr-auto navbar-nav"
        >
          <a
            class="nav-link"
            data-rb-event-key="/"
            href="/"
          >
            Home
          </a>
        </div>
      </div>
    </nav>
    <div
      class="container-fluid search-movie"
    >
      <div
        class="row d-flex align-items-center mt-4 mb-4"
      >
        <div
          class="col"
        >
          <h5>
            Searching
          </h5>
        </div>
        <div
          class="col d-flex justify-content-end"
        >
          <form
            class="form-inline"
          >
            <input
              class="form-control"
              type="text"
              value=""
            />
            <button
              class="btn btn-primary"
              data-testid="search-button"
            >
              Search
            </button>
          </form>
        </div>
      </div>
      <div
        class="row"
      />
    </div>
    <div
      class="container-fluid movie-app"
    >
      <div
        class="row d-flex mt-4 mb-4 ml-1"
      >
        <ul
          class="pagination"
        >
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="first-page-button"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                «
              </span>
              <span
                class="sr-only"
              >
                First
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="button-prev-page"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                ‹
              </span>
              <span
                class="sr-only"
              >
                Previous
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="button-page-2"
              href="#"
              role="button"
            >
              2
            </a>
          </li>
          <li
            class="page-item active"
          >
            <span
              class="page-link"
              data-testid="button-page-3"
            >
              3
              <span
                class="sr-only"
              >
                (current)
              </span>
            </span>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="button-page-4"
              href="#"
              role="button"
            >
              4
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="button-next-page"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                ›
              </span>
              <span
                class="sr-only"
              >
                Next
              </span>
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              class="page-link"
              data-testid="button-last-page"
              href="#"
              role="button"
            >
              <span
                aria-hidden="true"
              >
                »
              </span>
              <span
                class="sr-only"
              >
                Last
              </span>
            </a>
          </li>
        </ul>
        <form
          class=""
        >
          <div
            class="col"
          >
            <select
              class="form-control"
            >
              <option
                value="1"
              >
                1
              </option>
              <option
                value="2"
              >
                2
              </option>
              <option
                value="3"
              >
                3
              </option>
              <option
                value="4"
              >
                4
              </option>
              <option
                value="5"
              >
                5
              </option>
              <option
                value="6"
              >
                6
              </option>
              <option
                value="7"
              >
                7
              </option>
              <option
                value="8"
              >
                8
              </option>
              <option
                value="9"
              >
                9
              </option>
              <option
                value="10"
              >
                10
              </option>
            </select>
          </div>
        </form>
      </div>
    </div>
  </main>
</div>
`;
